### DISCUSSION TITLE
Compile error when using generic types in dependency injection

### DISCUSSION THREAD

**Discussion by HaoYang670** [tokio-rs/axum#3036] [Q&A]
### Summary

Hi, I meet some compile error when converting my server from concrete tyoe to generic type in  dependency injection.
I have a trait defined for interacting with database
```rust
#[trait_variant::make(HttpService: Send)]
pub trait DatabaseClient: Clone + Sized + Send + Sync + 'static {
// some async fn defined
}
```

And I define the client in State as a shared dependency
```rust
pub struct AppState<DB: DatabaseClient> {
    db: DB,
}
```

And I have a handler defined
```rust
async fn restart_engine<DB: DatabaseClient>(Path(id): Path<EngineId>, State(state): State<AppState<DB>>) -> Result<()> {...}
```

I have another function defined to get the router
```rust
fn get_engine_router<DB: DatabaseClient>() -> Router<AppState<DB>> {
    Router::new()
        .route("/:id/restart", post(restart_engine::<DB>))
}
```

but I get compile error: 
```
152 |         .route("/:id/restart", post(restart_engine::<DB>))
    |                                ---- ^^^^^^^^^^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for fn item `fn(Path<EngineId>, State<AppState<DB>>) -> impl Future<Output = Result<(), RucatError>> {restart_engine::<DB>}`
    |                                |
    |                                required by a bound introduced by this call
```

I have taken a look at https://github.com/tokio-rs/axum/blob/main/examples/dependency-injection/src/main.rs , but didn't find what may be wrong

### axum version

0.7.9

**Comment by HaoYang670** [comment]
It compiles after I replace `#[trait_variant::make(HttpService: Send)]` to `#[async_trait]` , maybe same as https://github.com/tokio-rs/axum/discussions/2718

**Reply by yanns** [reply]
I'm also using `#[async_trait]` fur such dependency injection.
